internal/databases/mongo/binary/backup.go

```diff
@@ -0,0 +1,319 @@
+	mongodMeta.TsStart = models.Timestamp{
+		TS:  backupCursorMeta.OplogStart.TS.T,
+		Inc: backupCursorMeta.OplogStart.TS.I,
+	}
+	mongodMeta.TsEnd = models.Timestamp{
+		TS:  backupCursorMeta.OplogEnd.TS.T,
+		Inc: backupCursorMeta.OplogEnd.TS.I,
+	}
+
+	return nil
+}
+
+func (backupService *BackupService) InitializeMongodBackupMeta(backupName string, permanent bool) error {
+	mongodVersion, err := backupService.MongodService.MongodVersion()
+	if err != nil {
+		return err
+	}
+	replSetName, err := backupService.MongodService.GetReplSetName()
+	if err != nil {
+		return err
+	}
+
+	userData, err := internal.GetSentinelUserData()
+	if err != nil {
+		return err
+	}
+
+	backupService.MongodBackupMeta = &MongodBackupMeta{
+		BackupName: backupName,
+		BackupType: backupType,
+
+		MongodMeta: MongodMeta{
+			Version:     mongodVersion,
+			ReplSetName: replSetName,
+		},
+
+		StartLocalTime: utility.TimeNowCrossPlatformLocal(),
+
+		Permanent: permanent,
+
+		BackupDirectories: []*BackupDirectoryMeta{},
+		BackupFiles:       []*BackupFileMeta{},
+
+		UserData: userData,
+	}
+
+	return nil
+}
+
+func (backupService *BackupService) AppendBackupFile(backupFile BackupCursorFile) error {
+	backupFilePath, err := backupService.LocalStorage.GetRelativeMongodPath(backupFile.FileName)
+	if err != nil {
+		return err
+	}
+
+	if previousBackupFileMeta, ok := backupService.BackupFilesMap[backupFilePath]; ok {
+		return processDoubleEncounterTheSameBackupFile(previousBackupFileMeta, backupFile)
+	}
+
+	backupMeta := backupService.MongodBackupMeta
+	directoryPath := filepath.Dir(backupFilePath)
+
+	for len(directoryPath) > 0 && directoryPath != "." {
+		if _, ok := backupService.BackupDirectoriesMap[directoryPath]; ok {
+			break
+		}
+
+		directoryStat, err := os.Stat(backupService.LocalStorage.GetAbsolutePath(directoryPath))
+		if err != nil {
+			return err
+		}
+
+		backupDirectoryMeta := BackupDirectoryMeta{
+			Path:     directoryPath,
+			FileMode: directoryStat.Mode(),
+		}
+		backupMeta.BackupDirectories = append(backupMeta.BackupDirectories, &backupDirectoryMeta)
+		backupService.BackupDirectoriesMap[directoryPath] = &backupDirectoryMeta
+
+		directoryPath = filepath.Dir(directoryPath)
+	}
+
+	backupFileStat, err := os.Stat(backupFile.FileName)
+	if err != nil {
+		return err
+	}
+
+	backupFileMeta := BackupFileMeta{
+		Path:             backupFilePath,
+		FileMode:         backupFileStat.Mode(),
+		Compression:      backupService.BackupStorage.GetCompression(),
+		UncompressedSize: backupFileStat.Size(), // backupFile.FileSize contains 0 for extended backup cursor
+	}
+	backupMeta.BackupFiles = append(backupMeta.BackupFiles, &backupFileMeta)
+	backupMeta.UncompressedDataSize += backupFile.FileSize
+	backupService.BackupFilesMap[backupFilePath] = &backupFileMeta
+
+	return nil
+}
+
+func processDoubleEncounterTheSameBackupFile(backupFileMeta *BackupFileMeta, backupFile BackupCursorFile) error {
```

Comment:
Do I correctly understand that we might encounter some files multiple times during the backup-push? When does it happen?